src/workers/python/PythonWorker.worker.ts

```diff
@@ -65,16 +64,19 @@ class PythonWorker extends Backend {
     }
 
     override async _runCodeInternal(code: string): Promise<any> {
-        await this.pyodide.loadPackagesFromImports(code);
         if (this.initialized) {
             // run the code, potentially polluting the namespace
             // Functions and variables defined by the user become global
             // We need them to be global to let doctest work out of the box
-            const result = this.pyodide.runPython(code);
+            if (await this.pyodide.globals.get(PROCESS_PYTHON_CODE)(code, false)) {
+                await this.pyodide.loadPackagesFromImports(code);
```

Comment:
I had a similar issue with `find_imports` on invalid syntax code, which led to a fix from pyodide: https://github.com/pyodide/pyodide/pull/1819

However it doesn't seem to actually be released which is weird: https://github.com/pyodide/pyodide/pull/1820

Anyway, I think it'd be better to instead wrap `loadPackagesFromImports` in a `try/catch` because if anything goes wrong it's probably not worth breaking the whole application anyway for the sake of automatic import installing. That's what I do at least.

Reply:
I do want friendly_traceback to intercept the potential syntax errors, which is why I run it through compile first. The pyodide package-loading code builds the AST to find import statements, so in order to catch these compile errors I need to do this. Unless you mean moving this part into Python aswell, which might not be a bad idea? :p

Reply:
Doesn't `eval_code_async` raise the SyntaxError so you can still handle it there?

Reply:
eval_code_async would also raise an Error, but I don't want it to try running the code as the imports wouldn't be available yet, so the raised Error could be the wrong one. Just compiling also sounds more efficient and clear.

Reply:
I don't think you understand. I'm suggesting this:

```js
try {
  await this.pyodide.loadPackagesFromImports(code);
} catch (e) {
  // Any failure to load packages shouldn't break everything
  console.error(e);
}
await this.pyodide.globals.get(PROCESS_PYTHON_CODE)(code);
```

Then you can also remove the boolean `run` argument and the associated conditional.

Reply:
This does not allow me to have friendly_traceback take care of the syntax error that could occur during importing. I know loadPackagesFromImports does a call to compile, but I want students to be able to understand where it went wrong. I believe loadPackages does not load any modules if a Syntax error occurred. However, in actual code, I think it is possible for the Syntax error to occur after a module was loaded and used. Would eval_code_async not give a MdouleNotFoundError instead of the Syntax error? This is why I want the compilation step to happen beforehand.

Reply:
If `code` contains a syntax error, `loadPackagesFromImports` will fail to load anything, and `eval_code_async` will raise the syntax error for friendly_traceback to catch.

If `code` doesn't contain a syntax error, `loadPackagesFromImports` might load something, and `eval_code_async` will not raise the syntax error for `code` itself. It might raise a syntax error when doing the actual import if the imported code contains invalid syntax, but you will still catch that as usual. It won't raise ModuleNotFoundError but I don't think that matters.

Running `loadPackagesFromImports` doesn't interfere with anything. The code will be compiled/parsed twice regardless of whether there's an error.